src/protocol/codecs/payload.rs

```diff
@@ -0,0 +1,93 @@
+use std::io;
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+use tracing::Span;
+
+use crate::protocol::{
+    codecs::{
+        tagmsg::Tag,
+        topic::{Topic, TopicCodec},
+    },
+    invalid_data,
+};
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [MsgOfInterest] contains a tag list in which the node is interested.
+#[derive(Debug)]
+pub struct MsgOfInterest {
+    /// Message tags for which the node is interested (subscribed).
+    // TODO(Rqnsom): convert to a hashset.
+    pub tags: String,
+}
+
+impl MsgOfInterest {
+    /// Constructs a [MsgOfInterest] payload from the [topics] list.
+    pub fn construct(mut topics: Vec<Topic>) -> Result<Self, io::Error> {
+        if topics.len() != 1 {
+            return Err(invalid_data!("expected a single topic"));
+        }
+
+        let tag_topic = topics.pop().unwrap();
+        if tag_topic.key != "tags" {
+            return Err(invalid_data!("expected 'tags' topic"));
+        }
+
+        Ok(Self {
+            tags: tag_topic.value,
+        })
+    }
+}
+
+/// [PayloadCodec] decodes the Algod message payload using a provided tag.
+#[derive(Clone)]
+pub struct PayloadCodec {
+    /// The associated node's span.
+    span: Span,
+
+    /// Represents a message payload type identifier.
+    pub tag: Option<Tag>,
+
+    /// Codec for topics which are key-value string pairs.
+    topic: TopicCodec,
+}
+
+impl PayloadCodec {
+    pub fn new(span: Span) -> Self {
+        Self {
+            span,
+            tag: None,
+            topic: TopicCodec::default(),
+        }
+    }
+}
+
+impl Decoder for PayloadCodec {
+    type Item = Payload;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        let tag = self
+            .tag
+            .take()
+            .ok_or_else(|| invalid_data!("tag not set"))?;
```

Comment:
I wonder if we shouldn't instead return early at the point where the `tag` is set :thinking:; that might obviate the option too

Reply:
alternatively, if the option is used only so that the tag can be removed once used, we should be able to assert here that the tag had already been set at this point

Reply:
> alternatively, if the option is used only so that the tag can be removed once used, we should be able to assert here that the tag had already been set at this point

Does that mean we can extract the tag with the `expect` function message? Yeah, we can do that.

I would prefer to keep the `option` solution just because it gives you more context when you read this function without knowing the fact that someone needs to set it each time before calling the `decode` method.
When I see that we take the tag out, I immediately know someone had set it before, and then we can assume it's the user of this codec (it automatically gives you a better understanding of how this works, sort of 😀).